[MRG] Adds KNNImputer

thomasjpfan opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
Continues https://github.com/scikit-learn/scikit-learn/pull/9348, and a part of https://github.com/scikit-learn/scikit-learn/pull/9212

Closes #2989
Closes #9348
Closes #9212 

#### What does this implement/fix? Explain your changes.
This PR cleans up the work done on https://github.com/scikit-learn/scikit-learn/pull/9348 and https://github.com/scikit-learn/scikit-learn/pull/9212:

1. Adds more tests.
2. Cleans up warnings/errors.
3. Currently, the distance between two completely nan vectors is nan. This causes the diagonal to sometimes be nan. This behavior can be changed depending on our preferences.

**Edit**: `KNNImputer` has been added this to PR with the following updates:

1. Address comments in https://github.com/scikit-learn/scikit-learn/pull/9212
2. Reduces the number of variables.
3. Memory optimizations.
4. Raises as soon as possible (before calculations).

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

ashimb9 committed:
Addressed review comments #5

ashimb9 committed:
Edited comments

ashimb9 committed:
Merge branch 'naneuclid' into knnimpute

ashimb9 committed:
KNN Imputation with masked_euclidean and sklearn.neighbors

ashimb9 committed:
fixed array base check

ashimb9 committed:
Fix column mean to nanmean

ashimb9 committed:
Added weight support and cleaned the code

ashimb9 committed:
Added inf check

ashimb9 committed:
Changed error message

ashimb9 committed:
Added test suite and example. Expanded docstring description

ashimb9 committed:
Changes to preprocessing __init__

ashimb9 committed:
Added KNNImputer exception for NaN and inf in estimator_checks

ashimb9 committed:
Moved _check_weights() to fit()

ashimb9 committed:
Addressed review comments - 1

ashimb9 committed:
Make NearestNeighbor import local to fit

ashimb9 committed:
Updated doc/modules/preprocessing.rst

ashimb9 committed:
More circular import fixes

ashimb9 committed:
pep8 fixes

ashimb9 committed:
Minor comment updates

ashimb9 committed:
Addressed review comments (part 2)

ashimb9 committed:
Fixed pyflex issues

ashimb9 committed:
Added test for callable weights and updated comments.

ashimb9 committed:
Pep8 fixes

ashimb9 committed:
Comment, doc, and pep8 fixes

ashimb9 committed:
Docstring changes

ashimb9 committed:
Changes to unit tests as per review comments

ashimb9 committed:
Tests moved to test_imputation

ashimb9 committed:
Addressed review comments

ashimb9 committed:
test changes

ashimb9 committed:
Test changes part 2

ashimb9 committed:
Fixed weight matrix shape issue

ashimb9 committed:
Minor changes

ashimb9 committed:
Fixed degenerate donor issue. Added tests

ashimb9 committed:
Further test updates

ashimb9 committed:
minor test fix

ashimb9 committed:
more minor changes

ashimb9 committed:
Moved weight_matrix inside if-weighted block

ashimb9 committed:
Addressed Review Comments

ashimb9 committed:
Fixed plot_missing example

ashimb9 committed:
Fixed Error Msg

ashimb9 committed:
Modified missing check for sparse matrix

ashimb9 committed:
Test update

ashimb9 committed:
Fixed nan check on sparse

ashimb9 committed:
Review Comments Addressed (partial)

ashimb9 committed:
Fix merge conflit

ashimb9 committed:
Updated doc module

ashimb9 committed:
Added support for using only neighbors with non-missing features

ashimb9 committed:
Test update

ashimb9 committed:
Import Numpy code for np.unique for older versions

ashimb9 committed:
Remove version check

ashimb9 committed:
Minor fix

ashimb9 committed:
Added strategy to only use neighbors with non-nan value

ashimb9 committed:
Sync with upstream and merge with master

ashimb9 committed:
Edit import path in test file

ashimb9 committed:
Error fixes with imports and examples

ashimb9 committed:
Added use_complete docstring

ashimb9 committed:
Changed comments and fixed docstring

ashimb9 committed:
Added more doctest fix and min neighbor check

ashimb9 committed:
fix docs

ashimb9 committed:
Increase col_max_missing threshold for example plot

ashimb9 committed:
Lower missing rate in demo since tests are failing

ashimb9 committed:
Remove redundant check and changes in plot

ashimb9 committed:
Handling insufficient neighbors scenario

ashimb9 committed:
Removed k actual neighbors algo

ashimb9 committed:
Addressed Comments

ashimb9 committed:
Merge branch 'master' into knnimpute

ashimb9 committed:
Sync with upstream and merge

ashimb9 committed:
Sync and merge

ashimb9 committed:
Minor bug fixes

ashimb9 committed:
Removing flotsam

ashimb9 committed:
Minor bug fixes

ashimb9 committed:
Merge to upstream

jnothman committed:
Revert changes to sklearn/neighbors

jnothman committed:
Merge branch 'master' into knnimpute

jnothman committed:
Revert changes to deprecated file

jnothman committed:
COSMIT _MASKED_METRICS -> _NAN_METRICS

jnothman committed:
'NaN' no longer stands for NaN

jnothman committed:
Fix missing_values validation

jnothman committed:
Attempt to reinstate neighbors changes

jnothman committed:
Fix up test failures

jnothman committed:
Fix flake8 issues in example

jnothman committed:
Default force_all_finite to True rather than False

jnothman committed:
Fix example usage

jnothman committed:
Fix masked_euclidean testing in nearest neighbors

jnothman committed:
Fix missing_values in masked_euclidean_distances

jnothman committed:
Can't subtract list and set in Py2

thomasjpfan committed:
RFC: Reduce diffs

thomasjpfan committed:
WIP

thomasjpfan committed:
MRG: Reduce diffs

thomasjpfan committed:
TST: Fix

thomasjpfan committed:
DOC: Refactor

thomasjpfan committed:
WIP

thomasjpfan committed:
RFC: Adjustments

thomasjpfan committed:
ENH: Completes implementation

thomasjpfan committed:
ENH: Adds to __init__

thomasjpfan committed:
DOC: Adds whats_new

thomasjpfan committed:
DOC: Adds autosummary

thomasjpfan committed:
RFC: Minor

thomasjpfan committed:
DOC: Grammer

thomasjpfan committed:
TST: Increases coverage

thomasjpfan committed:
TST: Increases coverage

thomasjpfan committed:
RFC: Minor

thomasjpfan committed:
DOC: Adjust order

jnothman commented on sklearn/neighbors/base.py:
```diff
@@ -22,8 +22,10 @@
 from ..metrics import pairwise_distances_chunked
 from ..metrics.pairwise import PAIRWISE_DISTANCE_FUNCTIONS
 from ..utils import check_X_y, check_array, gen_even_slices
+from ..metrics.pairwise import _NAN_METRICS
```
We no longer need the changes to Nearest Neighbors for KNNImputer to work. Do you think they are otherwise valuable?

thomasjpfan replied:
I think there is minimum value in having Nearest Neighbors support nan metrics.

jnothman replied:
Not quite sure what you're proposing here.

thomasjpfan replied:
One motivation by including the "masked euclidean distance" metric was to add it as a metric to Nearest Neighbors. Now that these changes are not needed in Nearest Neighbors, I think it would be best to add the KNNImputer back into this PR.

jnothman replied:
Feel free to try merge the two PRs!

jnothman commented on sklearn/metrics/pairwise.py:
```diff
@@ -32,6 +33,15 @@
 
 
+# Get mask for missing values
+def _get_mask(X, value_to_mask):
```
should we move this to utils?

jnothman replied:
Yes, please

jnothman replied:
Please do move this to utils

thomasjpfan replied:
Refactoring `_get_mask` to `utils` would causes changes unrelated to `KNNImputer` and `nan_euclidean_distances`. I can open another PR for this refactor.

thomasjpfan replied:
On second thought, I'll refactor it here. The addition of `KNNImputer` motivates the refactor.

jnothman reviewed (commented):
bigger question: is "masked euclidean distance" the right term to use? Is there another term in the literature?

thomasjpfan committed:
RFC: Removes euclidean metric from neighbors

thomasjpfan committed:
RFC

thomasjpfan committed:
ENH: Adds support for sparse

thomasjpfan committed:
RFC: Removes unused

jnothman committed:
Merge branch 'master' into knnimpute

adrinjalali committed:
DOC Fix warnings in examples (#12654)

Fixe or suppresses many warnings in examples, including matplotlib complaining about the non-gui backend.

chkoar committed:
DOC Add an example of inductive clustering (#10852)

jnothman committed:
DOC credit multiple authors of new example

jnothman committed:
DOC fix plot_iris references after files renamed

jnothman committed:
MNT Remove accidentally added example

agamemnonc committed:
docstring fix X in predict/predict_proba (#13004)

jnothman committed:
MNT more informative warning in estimator_checks (#13002)

rth committed:
MAINT Pin numpy version 1.5.* for pypy (#13011)

* Pin numpy version for pypy

* Add comment

* Updated comment to wait for next pypy release

zjpoh committed:
DOC Label Spreading clumping factor must be in (0, 1)  (#13015)

nixphix committed:
FIX Parallelisation of decomposition/sparse_encode (#13005)

pierretallotte committed:
FIX Convert the negative indices to positive ones in ColumnTransformer._get_column_indices (#13017)

rth committed:
MAINT Fix PyPy CI with numpy 1.15 (#13018)

qinhanmin2014 committed:
DOC Move datasets.mldata_filename to deprecated section in classes.rst

samwaterbury committed:
FIX Fix shuffle not passed in MLP (#12582)

thomasjpfan committed:
[MRG] Configure lgtm.yml for CPP (#13044)

baluyotraf committed:
FIX float16 overflow on accumulator operations in StandardScaler (#13010)

xhan7279 committed:
TST Use random state to initialize MLPClassifier. (#12892)

qinhanmin2014 committed:
API Deprecate externals.six (#12916)

qinhanmin2014 committed:
DOC Remove outdated doc in KBinsDiscretizer (#13047)

qinhanmin2014 committed:
DOC Remove outdated doc in KBinsDiscretizer

See #13074

qinhanmin2014 committed:
EXA Improve example plot_svm_anova.py (#11731)

vishaalkapoor committed:
DOC Correct TF-IDF formula in TfidfTransformer comments. (#13054)

gvacaliuc committed:
FIX an issue w/ large sparse matrix indices in CountVectorizer (#11295)

qinhanmin2014 committed:
DOC More details about the attributes in MinMaxScaler (#13029)

jeremiedbb committed:
DOC Clean up the advanced installation doc to remove python < 3.5 parts (#13064)

zjpoh committed:
API NMF and non_negative_factorization have inconsistent default init (#12989)

glemaitre committed:
MAINT: pin flake8 to stable version (#13066)

qinhanmin2014 committed:
EXA: fix xlabel and ylabel in plot_cv_digits.py (#13067)

glemaitre committed:
MAINT: remove flake8 pinning in circle ci (#13071)

CatChenal committed:
DOC Adds an example to PatchExtractor (#12819)

* Finalizes fix for #12202 from abandonned PR by @parul-l

* Completes 12202 fix abandoned by @parul-l

* Completes 12202 fix abandoned by @parul-l

* Extends 12202 fix over feature_extraction/image.py

* Closes #12202; white space removal

* Closes #12202; white space removal2

* Closes #12202; 3.5 compliance; added >>> in docstring code.

* Closes #12202; indentation discrep.

* Closes #12202; indentation discrep.2

* Example output formating; @jnotham

* Example output formating; forgot flake8

* Closes #12202; Removed excessive indentation in docstring (#wimlds)

* Closes #12202; Fixed inconsistent indentation in docstring (#wimlds)

* Closes #12202 (#wimlds); intentation, v3.5 compliance

* Closes #12202 (#wimlds); Output format issue solved with addition of # doctest: +NORMALIZE_WHITESPACE +DONT_ACCEPT_BLANKLINE +ELLIPSIS for print statements.

* Closes #12202 (#wimlds); Output format issue solved with addition of # doctest: +NORMALIZE_WHITESPACE +DONT_ACCEPT_BLANKLINE for print statements.

* Closes #12202 (#wimlds); Testing doctest direc.: removed DONT_ACCEPT_BLANKLINE (@jnothmam, @reshamas)

* Closes #12202 (#wimlds); Removed blank lines in doctest example.

jeremiedbb committed:
[MRG] Use Scipy cython BLAS API instead of bundled CBLAS (#12732)

NicolasHug committed:
MNT Ignore PendingDeprecationWarnings of matrix subclass with pytest (#13076)

qinhanmin2014 committed:
MNT More clean up after we remove python < 3.5 (#13078)

surgan12 committed:
MNT remove __future__ imports (#12791)

glemaitre committed:
MNT do not call fit twice in TransformedTargetetRegressor (#11641)

glemaitre committed:
FIX add support for non numeric values in MissingIndicator (#13046)

qinhanmin2014 committed:
MNT redundant from __future__ import (#13079)

qinhanmin2014 committed:
CI install pillow in pypy job (#13081)

qinhanmin2014 committed:
MNT Remove utils.validation._shape_repr (#13083)

Following dropping python 2.X support

qinhanmin2014 committed:
CI install pillow in Travis cron job (#13080)

qinhanmin2014 committed:
MNT Remove utils.fixes.euler_gamma (#13082)

jeremiedbb committed:
MNT Update setup and travis to support OpenMP (#13053)

walk-to-work committed:
DOC change from 'means_prior' to  'mean_prior' in BayesianGaussianMixture (#13090)

esvhd committed:
FIX added assertion for ValueError when cv iterator is empty (#12961)

* added assertion for ValueError when cv iterator is empty

* test value error message

* change to use match= instead of message= for pytest.raises

* fix regex escape char for flake8

* Added test case for unmached cv result length

* fix travis-ci issue

* use global X, y and n_split=3

* use true global X, y

thomasjpfan committed:
[MRG] DOC Adds explicit reference to clang (#13093)

jnothman committed:
DOC Refer to ONNX (#13095)

* DOC Refer to ONNX

* DOC drop sklearn-pmml from related projects

thomasjpfan committed:
[MRG] DOC Adds _pairwise property to dev docs (#13094)

bharatr21 committed:
DOC Correct code example in doc/developers/contributing.rst (#13098)

thomasjpfan committed:
RFC Better variable names

thomasjpfan committed:
DOC Update version to 0.21

thomasjpfan committed:
RFC Address comments

thomasjpfan committed:
RFC: Updates metric name to nan_euclidean

thomasjpfan committed:
RFC Address comments

thomasjpfan committed:
RFC Lowers the number of variables

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into masked_euclidean_check

thomasjpfan committed:
RFC Address comments

thomasjpfan committed:
RFC Uses pytest

thomasjpfan committed:
STY Flake8

thomasjpfan committed:
RFC Lowers LOC

thomasjpfan committed:
RFC

thomasjpfan committed:
RFC Rename to statistic

thomasjpfan committed:
RFC Uses less memory for distance matrix

thomasjpfan committed:
STY Spelling

thomasjpfan committed:
REV take_along_axis not in numpy 1.11.0

thomasjpfan committed:
TST Checks nan and normal euclidean distance

thomasjpfan committed:
ENH Updates whats_new

thomasjpfan commented:
`KNNImputer` has been added this to PR with the following updates:

1. Address comments in https://github.com/scikit-learn/scikit-learn/pull/9212
2. Reduces the number of variables.
3. Memory optimizations.
4. Raises as soon as possible (before calculations).

thomasjpfan committed:
DOC English

thomasjpfan committed:
DOC Rewords _compute_impute doc

thomasjpfan committed:
RFC _impute directly imputes X

thomasjpfan committed:
RF

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into masked_euclidean_check

jnothman reviewed (commented):
some old bits of review I forgot to submit

jnothman commented:
What makes this WIP?

jnothman commented on sklearn/impute.py:
```diff
@@ -1265,3 +1271,281 @@ def fit_transform(self, X, y=None):
+      Hastie, Robert Tibshirani, David Botstein and Russ B. Altman, Missing
+      value estimation methods for DNA microarrays, BIOINFORMATICS Vol. 17
+      no. 6, 2001 Pages 520-525.
+
```
I think we wanted to note somewhere that a reference implementation in R behaves a bit differently, using less than K neighbours where some of the nearest neighbours have the target feature missing

thomasjpfan replied:
I am not familiar with the R implementation, is there a reference/link to it that discuss how it works?

jnothman commented on sklearn/metrics/tests/test_pairwise.py:
```diff
@@ -577,6 +586,145 @@ def test_euclidean_distances():
+    (np.array([[-1, 1], [-1, 0]]), np.sqrt(2), -1),
+    (np.array([[0, -1], [1, -1]]), np.sqrt(2), -1)
+])
+@pytest.mark.parametrize("sparse", ['csr', 'csc', None])
```
How about just doing `parameterize('format', [np.asarray, csr_matrix, csc_matrix])`?

jnothman commented on sklearn/tests/test_impute.py:
```diff
@@ -1030,6 +1033,593 @@ def test_missing_indicator_sparse_param(arr_type, missing_values,
+
+def test_knn_imputer_complete_features():
+
+    # Test with use_complete=True
```
What does this mean?

thomasjpfan replied:
This has been removed. (Must have been an old parameter)

jnothman reviewed (commented):
It will be a big effort to properly review these tests (even if I've reviewed them before). It's a lot of code.

thomasjpfan commented on sklearn/impute.py:
```diff
@@ -16,13 +16,19 @@
 from .utils.sparsefuncs import _get_median
```
https://github.com/scikit-learn/scikit-learn/pull/9212#discussion_r221446713 I can put it back.

thomasjpfan commented:
This is back to WIP because I think the docstrings are not up to par. I am rewording some parts of docstrings. It will be back to MRG today.

The parameters that reference columns and rows seems inconsistent with how we use `samples` and `features`. These parameters include, `row_max_missing` and `col_max_missing`.

thomasjpfan committed:
RF Address comments

thomasjpfan committed:
RFC Uses to matrix

thomasjpfan committed:
DOC Changes variable name to reference samples and features

thomasjpfan committed:
DOC Grammer

thomasjpfan commented:
This PR has been updated:

1. Address comments
2. `row_max_missing` has been renamed to `feature_max_missing`.
3. `col_max_missing` has been renamed to `sample_max_missing`.

thomasjpfan committed:
RFC Moves _get_mask to utils

thomasjpfan committed:
BUG Fix

thomasjpfan committed:
RFC Minimizes diff

thomasjpfan committed:
RFC Moves _get_mask to utils

NicolasHug commented on doc/whats_new/v0.21.rst:
```diff
@@ -134,6 +134,10 @@ Support for Python 3.4 and below has been officially dropped.
   <benlawson>`.
 
+- |Feature| Added the class, :class:`impute.KNNImputer`, to impute missing 
+  values using k-Nearest Neighbors. :issue:`2989` by 
```
This should be a PR number not an issue number.

Also don't be afraid to add a link to your own PR ;)

Also ~~the class,~~? (above)

NicolasHug commented on doc/modules/impute.rst:
```diff
@@ -174,6 +182,49 @@ References
+sample, the donors are selected such that they have the feature present and 
+they are one of the ``n_neighbors`` nearest neighbors.
+
+then the sample can potentially have multiple sets of
```
Remove "then"?

Also needs upper case

amueller replied:
Each sample?

NicolasHug commented on doc/whats_new/v0.21.rst:
```diff
@@ -207,6 +211,10 @@ Support for Python 3.4 and below has been officially dropped.
   BLAS shipped with scipy instead of the bundled BLAS. :issue:`12732` by
   :user:`Jérémie du Boisberranger <jeremiedbb>`
 
+- |Feature| Added the :func:`metrics.nan_euclidean_distances` metirc, which
```
metric

Please link to this PR

NicolasHug commented on examples/impute/plot_missing_values.py:
```diff
@@ -79,6 +83,13 @@ def get_results(dataset):
     mean_impute_scores = get_scores_for_imputer(imputer, X_missing, y_missing)
 
+    # Estimate the score after kNN-imputation of the missing values
+    knn_estimator = make_pipeline(
```
nitpick: `knn_estimator` might be mistaken for an actual `KNeighborsRegressor` ... maybe `knn_impute_estimator`?

banilo replied:
How about something like 'knnimpute_and_rf' ?

NicolasHug commented on sklearn/metrics/pairwise.py:
```diff
@@ -32,6 +34,14 @@
 
 
+# Get mask for missing values
+def _get_mask(X, value_to_mask):
```
Should you remove this function now and use `_get_missing_mask` instead?

NicolasHug commented on sklearn/metrics/pairwise.py:
```diff
@@ -259,6 +297,137 @@ def euclidean_distances(X, Y=None, Y_norm_squared=None, squared=False,
+    squared : boolean, optional (default=False)
+        Return squared Euclidean distances.
+
+    missing_values : np.nan or integer, optional (default=np.nan)
```
The `Imputer`class seems to accept the string `"NaN"` rather than `np.nan`. I would suggest staying consistent unless there's a good reason not to?

thomasjpfan replied:
I think we are moving away from using "NaN": https://github.com/scikit-learn/scikit-learn/pull/11211#issuecomment-395708181

`SimpleImputer` uses `np.nan` and `Imputer` is being depreciated.

NicolasHug commented on sklearn/metrics/pairwise.py:
```diff
@@ -259,6 +297,137 @@ def euclidean_distances(X, Y=None, Y_norm_squared=None, squared=False,
+    >>> from sklearn.metrics.pairwise import nan_euclidean_distances
+    >>> nan = float("NaN")
+    >>> X = [[0, 1], [1, nan]]
+    >>> # distance between rows of X
```
Nitpick: Maybe add the comments on the same line as the code if possible?

NicolasHug commented on sklearn/metrics/pairwise.py:
```diff
@@ -1032,16 +1202,17 @@ def distance_metrics():
+    'l1'                    metrics.pairwise.manhattan_distances
+    'l2'                    metrics.pairwise.euclidean_distances
+    'manhattan'             metrics.pairwise.manhattan_distances
+    'nan_euclidean'      metrics.pairwise.nan_euclidean_distances
```
needs  a few more spaces

NicolasHug commented on sklearn/tests/test_impute.py:
```diff
@@ -1030,6 +1033,591 @@ def test_missing_indicator_sparse_param(arr_type, missing_values,
+    ])
+
+    X_nan = np.array([
+        [1, np.nan, 1,      1,      1.],
```
I'm not sure how I feel about this kind of formatting.

It's clearly visually nicer, but it's a pain as soon as you need to change something as you have to adjust the white spaces.

Pep8 discourages the use of a similar pattern:

```
Yes:

x = 1
y = 2
long_variable = 3

No:

x             = 1
y             = 2
long_variable = 3

```

Let's wait the other's feedback

NicolasHug reviewed (commented):
Mostly superficial comments for now.

This looks pretty good in general!

banilo commented on examples/impute/plot_missing_values.py:
```diff
@@ -79,6 +83,13 @@ def get_results(dataset):
 
+    # Estimate the score after kNN-imputation of the missing values
+    knn_estimator = make_pipeline(
+        KNNImputer(missing_values=0, sample_max_missing=0.99),
```
General comment on the whole example: Who abouting indicating missingness as NaN in the entire example, to make it more educational. In some place, having 0 to indicate missingness and replacing by 0 may be confusing to some people.

jnothman replied:
Agreed that missing_values=0 in this example is a bit obscure... but it is a separate issue.

banilo commented on examples/impute/plot_missing_values.py:
```diff
@@ -8,6 +8,9 @@
 
+With ``KNNImputer``, missing values can be imputed using the weighted
+or unweighted mean of the desired number of nearest neighbors.
+
```
Explanation string here: instead of repeating 'round-robin' from the docstring of the imputer, perhaps replace by something more intuitively understandable, such as: column-by-column fashion

banilo commented on examples/impute/plot_missing_values.py:
```diff
@@ -8,6 +8,9 @@
 The median is a more robust estimator for data with high magnitude variables
```
rather than just high magnitude, one may also mention explicitly that median is more robust in the presence of outliers

thomasjpfan committed:
DOC Updates whats new

thomasjpfan committed:
DOC Updates name

thomasjpfan committed:
CLN Address comments

amueller commented:
fix merge conflicts? ;)

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into masked_euclidean

thomasjpfan committed:
STY flake8

thomasjpfan committed:
BUG Fix

thomasjpfan committed:
BUG Add allow_nan tag

thomasjpfan committed:
TST: Improves test coverage

thomasjpfan commented:
This PR has been updated:

1. Improves test coverage of `KNNImputer` for missing values other than `np.nan`.
2. Refactors `_get_mask` to `get_missing_mask` in `sklearn.utils.mask`.

banilo commented on sklearn/impute.py:
```diff
@@ -1279,3 +1269,286 @@ def fit_transform(self, X, y=None):
+
+        self.weights = _check_weights(self.weights)
+        self.fit_X_ = X[
+            mask.sum(axis=1) <= (mask.shape[1] * self.feature_max_missing)]
```
Simple is better than complex -> perhaps consider splitting into more than one line of code

banilo commented on sklearn/metrics/pairwise.py:
```diff
@@ -1020,6 +1181,7 @@ def chi2_kernel(X, Y=None, gamma=1.):
     'l1': manhattan_distances,
     'manhattan': manhattan_distances,
     'precomputed': None,  # HACK: precomputed is always allowed, never called
+    'nan_euclidean': nan_euclidean_distances,
```
, at the end of line seems superfluous

amueller commented on sklearn/metrics/pairwise.py:
```diff
@@ -1071,7 +1234,8 @@ def _parallel_pairwise(X, Y, func, n_jobs, **kwds):
     """Handle the callable case for pairwise_{distances,kernels}
     """
-    X, Y = check_pairwise_arrays(X, Y)
+    force_all_finite = False if callable(metric) else True
```
What's the reasoning behind this?

amueller reviewed (commented):
started with some initial comments

thomasjpfan commented on doc/modules/impute.rst:
```diff
@@ -174,6 +182,49 @@ References
+
+Each missing feature is then imputed as the average, either weighted or
+unweighted, of these neighbors. When the number of donor neighbors is less
+than ``n_neighbors``, the training set average for that feature is
```
When there are not enough donors, currently this imputer considers the whole training set as its donors.

Thinking about it now, this seems a little strange.

jnothman replied:
Well we don't need to handle the insufficient donors case specially if there's no weighting.  But whether we are right to disregard weighting when there are insufficient neighbours, I don't know.

thomasjpfan replied:
@amueller To clarify, what I meant by "everyone is a donor" is that this implementation uses the feature mean for the entire training set (ignoring the missing values) for imputation. This is done when there are not enough donors for a given sample.

jnothman replied:
I think it's clear, but if there aren't enough donors then at most n_neighbors available donors = everyone.

thomasjpfan commented on sklearn/impute.py:
```diff
@@ -1279,3 +1269,286 @@ def fit_transform(self, X, y=None):
+        self.weights = _check_weights(self.weights)
+        self.fit_X_ = X[
+            mask.sum(axis=1) <= (mask.shape[1] * self.feature_max_missing)]
+        self.statistics_ = np.ma.array(X, mask=mask).mean(axis=0).data
```
Maybe raise an error if one of the statistics is nan.

thomasjpfan commented on sklearn/impute.py:
```diff
@@ -1265,3 +1255,283 @@ def fit_transform(self, X, y=None):
+                        force_all_finite=force_all_finite, copy=self.copy)
+
+        # Check if sufficient neighboring samples available
+        if X.shape[0] < self.n_neighbors:
```
Maybe we do not need this check. If there are 5 observations and k=5, this imputer becomes a mean imputer. It still works, but it will not use any of the KNN concepts.

thomasjpfan reviewed (commented):
Couple of notes to myself.

jnothman commented:
What's new will need moving to 0.22

thomasjpfan committed:
CLN Address comments

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into masked_euclidean

thomasjpfan committed:
DOC Moves whats new

thomasjpfan committed:
CLN Address diffs

thomasjpfan committed:
DOC Update version

jnothman commented on doc/modules/impute.rst:
```diff
@@ -178,6 +186,49 @@ References
+Nearest neighbors imputation
+============================
+
+The :class:`KNNImputer` class provides imputation for completing missing
```
no, it just means "for replacing missing values" or something. It's unnecessarily verbose, I suppose.

jnothman reviewed (commented):
Please move this into sklearn/impute/_knn.py so we can see future review comments there.

amueller commented on sklearn/impute/_knn.py:
```diff
@@ -0,0 +1,290 @@
+        row_total_missing = mask.sum(axis=1)
+        bad_rows = row_total_missing > (mask.shape[1] *
+                                        self.feature_max_missing)
+        if np.any(bad_rows):
```
Is it faster to do this than to just compute them? Or is the point that they might all be NaN and then we get an error below?

amueller reviewed (commented):
Haven't gone through all details yet but most ;)
The defaults seem a bit arbitrary to me, and I'm not sure if sample_max_missing is a good parameter to have. It's inconsistent with the default behavior of SimpleImputer, which is quite inconvenient. What happens if in cross-validation some of the folds are below this? What's the user to do? Change the threshold?

thomasjpfan committed:
DOC Rewords

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into masked_euclidean

thomasjpfan committed:
CLN Move to _knn

thomasjpfan committed:
DOC Moves whats_new

thomasjpfan committed:
CLN Less diffs

thomasjpfan committed:
DOC Rewords user guide

thomasjpfan committed:
ENH Simplifies knn imputer

thomasjpfan committed:
ENH Address comments

thomasjpfan committed:
CLN Rename variable

thomasjpfan committed:
ENH Improves float32 handling

thomasjpfan committed:
DOC Adds comments

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into masked_euclidean

thomasjpfan commented:
This PR was updated:

1. Address doc comments (Removes donor wording)
2. Removes `sample_max_missing` and `feature_max_missing`
3. Adjusts `nan_euclidean_distance` to use `euclidean_distances` to getter handle float32

thomasjpfan committed:
BUG Makes fit_X private

amueller commented:
merge conflicts ;)

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into masked_euclidean

amueller commented:
still merge conflicts? Or maybe again lol.

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into masked_euclidean

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into masked_euclidean

amueller commented on sklearn/impute/_knn.py:
```diff
@@ -0,0 +1,264 @@
+    sample can be different depending on the particular feature being imputed.
+    When the number of available neighbors is less than ``n_neighbors``, the
+    training set average for that feature is used during imputation. When a
+    sample has more than a ``feature_max_missing`` fraction of its features
```
``feature_max_missing`` doesn't exist any more.

amueller replied:
If it's less than ``k_neighbors`` we're using the mean, right?

thomasjpfan replied:
Yes, I hope the sentence from the docstring is clear enough about this:

> When the number of available neighbors is less than ``n_neighbors``, the
    training set average for that feature is used during imputation.

amueller replied:
yeah that's fine.

amueller commented on sklearn/impute/_knn.py:
```diff
@@ -0,0 +1,264 @@
+    """Imputation for completing missing values using k-Nearest Neighbors.
+
+    Each sample's missing values are imputed using values from ``n_neighbors``
+    nearest neighbors found in the training set. Each missing feature is then
```
If a feature is always missing, we drop it. This is not documented.

amueller commented on sklearn/impute/_knn.py:
```diff
@@ -0,0 +1,264 @@
+    Attributes
+    ----------
+    statistics_ : array, shape (n_features, )
+        The 1-D array contains the mean of each feature calculated using
```
Contains ``np.NaN`` if a feature is always missing?

amueller commented on sklearn/impute/_knn.py:
```diff
@@ -0,0 +1,264 @@
+
+        Returns
+        -------
+        X : array-like, shape = (n_samples, n_features)
```
Output shape is incorrect if a feature is always missing.

amueller commented on sklearn/impute/_knn.py:
```diff
@@ -0,0 +1,264 @@
+
+    def _calc_impute(self, dist, receivers_idx, X_col, fit_X_col,
+                     non_missing_fix_X_col, statistic):
+        """Helper function to calculate the imputation value of a single
```
"to impute a single column"? so we can fit it in one line?

amueller commented on sklearn/impute/_knn.py:
```diff
@@ -0,0 +1,264 @@
+
+        Parameters
+        ----------
+        dist : array-like, shape = (n_receivers, n_train_samples, )
```
```suggestion
        dist : array-like, shape = (n_receivers, n_train_samples)
```

amueller commented on sklearn/impute/_knn.py:
```diff
@@ -0,0 +1,264 @@
+        non_missing_fix_X_col : array-like, shape = (n_train_samples, )
+            mask for missing values corresponding to ``fit_X_col``
+
+        statistic : float
```
Would it be easier to call this ``mean``?

thomasjpfan replied:
In the future, we may want to enable other statistics?

amueller replied:
alright

amueller commented on sklearn/impute/_knn.py:
```diff
@@ -0,0 +1,264 @@
+        fit_X_col : array-like, shape = (n_train_samples, )
+            column from training
+
+        non_missing_fix_X_col : array-like, shape = (n_train_samples, )
```
```suggestion
        non_missing_fit_X_col : array-like, shape = (n_train_samples, )
```

amueller commented on sklearn/impute/_knn.py:
```diff
@@ -0,0 +1,264 @@
+            mask for missing values corresponding to ``fit_X_col``
+
+        statistic : float
+            statistic for column generated from fitting ``fit_X_col``
```
Doesn't document ``Returns``.

amueller replied:
Does the shape/return type depend on the input? That's kinda funky

thomasjpfan replied:
Refactored this to be less funky.

amueller commented on sklearn/impute/_knn.py:
```diff
@@ -0,0 +1,264 @@
+
+        # Retrieve donor values and calculate kNN score
+        donors = fit_X_col[potential_donors_idx].take(donors_idx)
+        donors = np.ma.array(donors,
```
can we make this two lines?

amueller commented on sklearn/impute/_knn.py:
```diff
@@ -0,0 +1,290 @@
+
+    Attributes
+    ----------
+    statistics_ : array, shape (n_features, )
```
private now

amueller reviewed (commented):
Seems mostly good, though I don't understand the handling of sparse matrices and the docs have some minor issues.

thomasjpfan commented on doc/modules/impute.rst:
```diff
@@ -187,6 +195,49 @@ References
 
+.. _knnimpute:
+
+Nearest neighbors imputation
```
In the second sentence of the first paragraph it states:

>  By default, a eucliean distance
metric that supports missing values,
:func:`~sklearn.metrics.nan_euclidean_distances`, is used to find the
nearest neighbors.

thomasjpfan committed:
CLN Moves force_all_finite up

thomasjpfan commented on sklearn/metrics/pairwise.py:
```diff
@@ -281,6 +311,128 @@ def euclidean_distances(X, Y=None, Y_norm_squared=None, squared=False,
+                                 force_all_finite=force_all_finite, copy=copy)
+    # Get missing mask for X
+    if issparse(X):
+        missing_X = csr_matrix(
```
I agree this is not a good way to do sparse. This was a quick way to get sparse matrices to work with this metric. 

In its current form, `KNNImputer` does not support sparse matrices, so we can remove this hack and let `nan_euclidean_distances` not support sparse matrices.

amueller replied:
Fine by me.

thomasjpfan committed:
CLN Address comments

thomasjpfan committed:
REV Remove missing test

thomasjpfan commented on sklearn/metrics/pairwise.py:
```diff
@@ -1533,6 +1692,15 @@ def pairwise_distances(X, Y=None, metric="euclidean", n_jobs=None, **kwds):
+        if np.all(_get_missing_mask(X.data if issparse(X)
+                                    else np.array(X), missing_values)):
+            raise ValueError(
+                "One or more samples(s) only have missing values.")
```
Removed this check. Our `nan_euclidean_distances` can handle all samples with all missing values. (It has `nan` distances to everyone).

Overall, I do not think its `pairwise_distances` job to check this.

amueller replied:
I agree.

amueller commented on sklearn/impute/_knn.py:
```diff
@@ -0,0 +1,270 @@
+
+        Returns
+        -------
+        imputed_values: array, shape=(n_receivers, )
```
```suggestion
        imputed_values: array, shape=(n_receivers,)
```

amueller commented on sklearn/impute/_knn.py:
```diff
@@ -0,0 +1,270 @@
+
+            receivers_idx = np.where(mask[:, col])[0]
+
+            # Row index for receivers and potential donors (pdonors)
```
```suggestion
            # Row index for receivers and potential donors
```

thomasjpfan committed:
ENH Removes sparse support

thomasjpfan committed:
STY Flake8

thomasjpfan committed:
ENH Moves all _get_mask to utils

amueller committed:
Merge branch 'master' into masked_euclidean

amueller commented on sklearn/metrics/pairwise.py:
```diff
@@ -281,6 +313,116 @@ def euclidean_distances(X, Y=None, Y_norm_squared=None, squared=False,
+    # Adjust distances for missing values
+    XX = X.multiply(X) if issparse(X) else X * X
+    YY = Y.multiply(Y) if issparse(Y) else Y * Y
+    distances -= safe_sparse_dot(XX, missing_Y.T, dense_output=False)
```
distances is dense in any case, right? and the RHS can be sparse? I'm maybe a bit confused by using ``-=`` between a dense and sparse matrix but if it works it seems fine.

thomasjpfan replied:
Since sparse support was removed, this has been converted to use `np.dot`.

thomasjpfan commented on sklearn/utils/mask.py:
```diff
@@ -0,0 +1,21 @@
+from .fixes import _object_dtype_isnan
+
+
+def _get_mask(X, value_to_mask):
```
This does not support sparse.

thomasjpfan committed:
CLN Completely remove sparse support

amueller reviewed (approved):
Looks good now, I think.
Would be great to get another review, but I think @jnothman is in release mode?

jnothman commented:
> @jnothman is in release mode?

I suppose so, though apart from being time-sensitive, release is not a big deal. Would be good to have it out of the way...

I find myself rather tired of this PR having already reviewed earlier recensions so many times!! But I'll try to give it another look.

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into masked_euclidean

amueller commented:
@jnothman yes, I think I only have reviewed a fraction of the versions of this PR that you have (though also plenty), so I get that. But I think it's finally somewhere good now ;)

jnothman commented on doc/modules/impute.rst:
```diff
@@ -32,6 +32,13 @@ missing values (e.g. :class:`impute.IterativeImputer`).
 Univariate feature imputation
 =============================
 
+Imputer transformers can be used to create pipelines that support data with
```
This doesn't seem pertinent to the heading.

jnothman commented on doc/modules/impute.rst:
```diff
@@ -32,6 +32,13 @@ missing values (e.g. :class:`impute.IterativeImputer`).
+
+
+Simple univariate imputation
+============================
```
This seems to be a duplication of the preceding heading??

jnothman commented on doc/modules/impute.rst:
```diff
@@ -180,6 +187,42 @@ References
+values using the k-Nearest Neighbors approach. By default, a eucliean distance
+metric that supports missing values,
+:func:`~sklearn.metrics.nan_euclidean_distances`, is used to find the
+nearest neighbors. Each missing feature are 
```
```suggestion
nearest neighbors. Each missing feature is
```

jnothman commented on doc/modules/impute.rst:
```diff
@@ -180,6 +187,42 @@ References
+nearest neighbors. Each missing feature are 
+imputed using values from ``n_neighbors`` nearest neighbors that have a value
+for the feature. The feature of the neighbors are averaged uniformly or
+weighted by distance. When the number of neighbors is less than 
```
Possibly:
```suggestion
weighted by nearness. When the number of neighbors is less than 
```

thomasjpfan replied:
How about:

> The feature of the neighbors are averaged uniformly or weighted by distance to each neighbor

jnothman commented on doc/modules/impute.rst:
```diff
@@ -180,6 +187,42 @@ References
+imputed using values from ``n_neighbors`` nearest neighbors that have a value
+for the feature. The feature of the neighbors are averaged uniformly or
+weighted by distance. When the number of neighbors is less than 
+``n_neighbors``, the training set average for that feature is used for
```
This is not really a special case if the regression is unweighted. If it's weighted, I'm not sure why the mean of all training values is a good choice.

thomasjpfan replied:
For the weighted case, we can weight this mean by distances as well.

amueller replied:
For the unweighted case: I might have argued this point wrongly before? It could be that the feature is present in a lot of samples, but none of them have features in common with the current sample, so they are not considered neighbors. So the mean includes samples that are not neighbors.

For the weighted case: I still think it's reasonable there.

jnothman commented on sklearn/impute/_knn.py:
```diff
@@ -0,0 +1,270 @@
+    When the number of available neighbors is less than `n_neighbors`, the
+    training set average for that feature is used during imputation. If a
+    feature is always missing, it is removed during `transform`.
+
```
Read more in the User Guide

jnothman commented on sklearn/impute/_knn.py:
```diff
@@ -0,0 +1,270 @@
+    Each sample's missing values are imputed using values from `n_neighbors`
+    nearest neighbors found in the training set. Each missing feature is then
+    imputed as the average, either weighted or unweighted, of these neighbors.
+    If a sample has more than one feature missing, then the neighbors for that
```
I think this note is currently missing from the user guide. I would expect the guide to be more informative on such details than the description in the docstring.

jnothman commented on sklearn/impute/_knn.py:
```diff
@@ -0,0 +1,270 @@
+        - callable : a user-defined function which accepts an
+          array of distances, and returns an array of the same shape
+          containing the weights.
+
```
fwiw we could also provide median and mode weightings...

thomasjpfan replied:
This can be done in another PR. For now our `_get_weight` code base only support `uniform`, `distance` and callables.

jnothman replied:
Yeah, I've realised that you can't calculate median and mode as weightings... they depend on the value, not just the weight.

jnothman commented on sklearn/impute/_knn.py:
```diff
@@ -0,0 +1,270 @@
+        neighbors search because they have too many missing features or
+        all of the sample's k-nearest neighbors (i.e., the potential donors)
+        also have the relevant feature value missing. If a feature is always
+        missing, the `mean_` will be `np.nan`, andt that feature will be
```
```suggestion
        missing, the `mean_` will be `np.nan`, and that feature will be
```

jnothman reviewed (commented):
a start

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into masked_euclidean

thomasjpfan committed:
CLN Address comments

jnothman committed:
Remove statistics_ from KNNImputer (#8)

TODO: add tests with weighting and <n_neighbors donors

thomasjpfan committed:
Revert "Remove statistics_ from KNNImputer (#8)"

This reverts commit 6dcef51f6b9a062fe16874f7cdd975ac05e4cfbf.

thomasjpfan committed:
TST Adds more tests

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into masked_euclidean

thomasjpfan committed:
TST Adds test on dropping features

thomasjpfan commented:
Moving discussion https://github.com/thomasjpfan/scikit-learn/pull/8 here:

https://github.com/thomasjpfan/scikit-learn/pull/8#issuecomment-517566875

Consider n_neighbors=2 and where some distances are not defined (nan):

```py
X = np.array([[na, 2], [na, 1], [3, na]])
```
The mean would not be able to be weighted by distance.

> In what context are distances NaN? Surely the user is responsible for choosing a weight function that corresponds to the possible distances???

The distance between the `[3, na]` and the other samples is `nan` because they do not have matching features.

https://github.com/thomasjpfan/scikit-learn/pull/8#issuecomment-517569833

> Is the problem only really that we need to define what happens with these NaN distances for weights='uniform' and weights='distance' (give it a weight of eps such that its value is only taken into account if no donors have finite weight?) and tell the user that custom weight functions should handle them too?

It is possible to do this.

> Yes, I suppose just always resorting to means is a simpler solution... it just means that you're going to get big behaviour change with weighted regression as n_neighbors / n_samples changes.

> And I would still suggest that precomputing and storing statistics_ for an edge case is premature optimisation.

jnothman commented:
Just to reword what I said above: While I think this is an edge case and we
should probably not optimise it too much, having distance weighting should
mean that once n_neighbors is sufficient to include the local neighborhood
of each sample, increasing it further should have negligible impact.
Resorting to the training set mean means that increasing n_neighbors is not
smooth at all.

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into masked_euclidean

thomasjpfan committed:
WIP

jnothman commented:
So what happens when all points have no distance? we still resort to the
unweighted mean? I suppose that's fine.

thomasjpfan committed:
ENH Adjusts handling when there is not enough neighbors

thomasjpfan commented:
> So what happens when all points have no distance? we still resort to the
unweighted mean? I suppose that's fine.

Yup.

This PR has been updated:

1. If all points have no distance, we still resort to the unweighted mean.
2. Removes `statistics_`
3. For a given feature, when there are not enough neighbors and there exists at least one neighbor, those neighbors will be use for imputation (weighted or unweighted).
4. Updated docs to reflect this change.

jnothman commented:
So then it's not entirely satisfying that one neighbour with finite
distance will be the only donor when there are other neighbours of NaN
distance. You could further engineer the solution to take the training set
mean when there is only one not zero finite distance among the donors. But
I'm okay with what you're proposing and I think going beyond that is over
engineering. I think we are right to make distance weighting smooth under
increasing n_neighbours.

We could leave a note that this extreme case behaviour is experimental and
subject to change.

jnothman commented:
We would be better off using our effort to make it possible to impute with
medians or modes.

amueller commented:
Or one more of 1000 other things tbh ;)
I'll try to do another review today or tomorrow. @jnothman do you approve? Do you want to check the details? I feel like this PR should end soon ;)

jnothman commented:
@amueller I will try find time for reviewing but there's a lot of follow
just to keep on top of things.

amueller commented:
@jnothman just do what I did and give up, it makes life so much easier ;)

jnothman commented on sklearn/impute/_knn.py:
```diff
@@ -0,0 +1,271 @@
+        # Pairwise distances between receivers and fitted samples
+        dist = pairwise_distances(X[row_missing_idx, :], self._fit_X,
+                                  metric=self.metric,
+                                  squared=False,
```
This will break for an arbitrary callable metric, won't it?

thomasjpfan replied:
If you are referring to `force_all_finite`, this has been updated to `force_all_finite` which is defined in reference to `self.missing_value`.

If you are referring to `missing_value`, the `metric` docstring specifies that `missing_values` need to be a keyword.

amueller replied:
Thought he was referring to "squared"?

thomasjpfan replied:
Ah yes. PR updated to remove "squared".

jnothman commented on doc/modules/impute.rst:
```diff
@@ -180,6 +180,47 @@ References
+training set average for that feature is used during imputation. If there is at
+least one neighbor with a defined distance, the weighted or unweighted average
+of the remaining neighbors will be used during imputation. If a feature is
+always missing, it is removed during `transform`.For more information on the
```
"always missing" -> "always missing in training"

amueller replied:
```suggestion
always missing, it is removed during `transform`. For more information on the
```

jnothman commented on sklearn/impute/_knn.py:
```diff
@@ -0,0 +1,271 @@
+class KNNImputer(TransformerMixin, BaseEstimator):
+    """Imputation for completing missing values using k-Nearest Neighbors.
+
+    Each sample's missing values are imputed using values from `n_neighbors`
```
"using values" -> "using the mean value". I'd be happy to remove all sentences after this one. Indeed, if there were to be another sentence, I'd choose to mention how distance is calculated, basically "Two samples are close if the features that neither is missing are close."

jnothman reviewed (commented):
Still mid-review

amueller commented on sklearn/impute/_knn.py:
```diff
@@ -0,0 +1,271 @@
+    If a sample has more than one feature missing, then the neighbors for that
+    sample can be different depending on the particular feature being imputed.
+    When the number of available neighbors is less than `n_neighbors` and
+    there are no defined distances to the training set, the training set
```
"no defined distances" is a bit unclear. And as @jnothman said, I feel this level of detail should be in the user guide.

amueller reviewed (approved):
the new changes look good. The docs could use some slight refactoring and polish but I'd love to get this in soon ;)

thomasjpfan committed:
CLN Address comments

thomasjpfan committed:
CLN Address comments

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into masked_euclidean

thomasjpfan committed:
ENH Updates check_is_fitted

thomasjpfan committed:
CLN Removes squared

jnothman reviewed (commented):
This is another comment I had pending. Not sure when I'll give it another full pass

thomasjpfan commented on sklearn/impute/tests/test_knn.py:
```diff
@@ -0,0 +1,625 @@
+
+    # Test with "callable" weight
+    def no_weight(dist):
+        return None
```
When `weights` is a callable, `_get_weights` returns `None`:

https://github.com/scikit-learn/scikit-learn/blob/2157880f6580584dd3f3b5a438f81824170a7610/sklearn/neighbors/base.py#L99-L100

thomasjpfan committed:
CLN Address jnothman's comment

jnothman commented on sklearn/impute/_knn.py:
```diff
@@ -0,0 +1,262 @@
+        return X[:, valid_idx]
+
+    def _more_tags(self):
+        return {'allow_nan': True}
```
```suggestion
        return {'allow_nan': is_scalar_nan(self.missing_values)}
```

jnothman commented on sklearn/impute/_knn.py:
```diff
@@ -0,0 +1,262 @@
+
+        Parameters
+        ----------
+        dist_pot_donors : array-like, shape=(n_receivers, n_train_samples)
```
Should reformat according to the recent edicts on these matters.

jnothman reviewed (commented):
@amueller, this one doesn't store any attributes ending in `_`.

Sorry, still not really able to give this a full review.

thomasjpfan commented:
@jnothman `check_is_fitted` checks for `_` in the beginning as well:

https://github.com/scikit-learn/scikit-learn/blob/35c0ca0325fa204759e040fb6236ae24f648b046/sklearn/utils/validation.py#L927

thomasjpfan committed:
CLN Refactor and improve docstring

thomasjpfan committed:
CLN Address joels comments

amueller commented:
tests are failing...

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into masked_euclidean

thomasjpfan commented:
Test failure was unrelated, now that https://github.com/scikit-learn/scikit-learn/pull/14689 is merged, merging this with master should fix it.

jnothman commented on sklearn/impute/tests/test_knn.py:
```diff
@@ -0,0 +1,638 @@
+    dist = nan_euclidean_distances(X)
+    r1c1_nbor_dists = dist[1, [0, 2, 3, 4, 5]]
+    r1c3_nbor_dists = dist[1, [0, 3, 4, 5, 6]]
+    r1c1_nbor_wt = (1 / r1c1_nbor_dists)
```
```suggestion
    r1c1_nbor_wt = 1 / r1c1_nbor_dists
```

jnothman commented on sklearn/impute/tests/test_knn.py:
```diff
@@ -0,0 +1,638 @@
+    r1c1_nbor_dists = dist[1, [0, 2, 3, 4, 5]]
+    r1c3_nbor_dists = dist[1, [0, 3, 4, 5, 6]]
+    r1c1_nbor_wt = (1 / r1c1_nbor_dists)
+    r1c3_nbor_wt = (1 / r1c3_nbor_dists)
```
```suggestion
    r1c3_nbor_wt = 1 / r1c3_nbor_dists
```

jnothman commented on sklearn/impute/tests/test_knn.py:
```diff
@@ -0,0 +1,638 @@
+
+
+@pytest.mark.parametrize("na", [-1, np.nan])
+def test_knn_imputer_with_weighted_features(na):
```
"weighted features" seems to be a strange name. This also appears to overlap in purpose with `test_knn_imputer_weight_distance`. What's the distinction?

thomasjpfan replied:
There was not a distinction. `test_knn_imputer_with_weighted_features` was removed and the tests from there was moved into `test_knn_imputer_weight_distance`

jnothman reviewed (commented):
Otherwise LGTM!!!

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into masked_euclidean

thomasjpfan committed:
STY Fix

thomasjpfan committed:
CLN Combines weighted tests

thomasjpfan committed:
BUG Fixes bug with test without missing values

thomasjpfan commented on doc/modules/impute.rst:
```diff
@@ -180,6 +180,47 @@ References
+least one neighbor with a defined distance, the weighted or unweighted average
+of the remaining neighbors will be used during imputation. If a feature is
+always missing in training, it is removed during `transform`. For more
+information on the methodology, see ref. [OL2001]_.
```
It looks like they use distance weighting by default.

thomasjpfan replied:
Gerhard Tutz, Shahla Ramzan, 
[Improved Methods for the Imputation of Missing Data by Nearest Neighbor Methods](https://epub.ub.uni-muenchen.de/21722/1/tr172.pdf) shows in their experiments that weighted kNN performs better than unweighted.

thomasjpfan replied:
Lorenzo Beretta* and Alessandro Santaniello, [Nearest neighbor imputation algorithms: a critical evaluation](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4959387/pdf/12911_2016_Article_318.pdf), shows that weighted is slightly better than unweighted.

thomasjpfan committed:
STY Fix

thomasjpfan committed:
ENH Stores fit_X mask during fit

thomasjpfan commented on sklearn/impute/tests/test_knn.py:
```diff
@@ -0,0 +1,638 @@
+    ])
+    knn = KNNImputer(missing_values=na, n_neighbors=2).fit(X)
+
+    X_transform = knn.transform(X)
```
PR updated with this suggestion.

thomasjpfan commented:
The mask of `fit_X` is now stored as a private attribute during `fit`.

jnothman reviewed (approved):
I think this is good to hit the road.

Congrats @ashimb9 and thanks @thomasjpfan for finishing it off!

amueller commented:
yayyyy finally ;)

banilo commented:
Glad to see this one in, too!

Danilo



> Am 04.09.2019 um 6:43 PM schrieb Andreas Mueller <notifications@github.com>:
> 
> yayyyy finally ;)
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/scikit-learn/scikit-learn/pull/12852?email_source=notifications&email_token=AA5Z6RABX22LODMU7SHEIDDQH7QSZA5CNFSM4GL4N2AKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD54GS7A#issuecomment-527985020>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AA5Z6RB4ZL7CGTRMJDWPFODQH7QSZANCNFSM4GL4N2AA>.
>

chitcode commented:
Very happy to see this.
